nodedb_test.go

```diff
@@ -248,3 +244,128 @@ func makeHashes(b *testing.B, seed int64) [][]byte {
+
+	ndb, _, _, _ := createNodeDbWithMocks(ctrl)
+	err := ndb.SaveFastNode(mockNode)
+	require.Error(t, err)
+}
+
+// TODO: We should have explicit argument types for below functions
+func TestNodeDB_SaveFastNode_ReturnErrorWhenNdbBatchFails(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockNode := NewMockNode(ctrl)
+	ndb, _, mockBatch, _ := createNodeDbWithMocks(ctrl)
+
+	var buf bytes.Buffer
+	buf.Grow(40)
+
+	mockNode.EXPECT().getKey().Return([]byte("k1")).Times(2)
+	mockNode.EXPECT().EncodedSize().Return(1).Times(1)
+	mockNode.EXPECT().WriteBytes(gomock.AssignableToTypeOf(&buf)).
+		Return(nil).
+		Do(func(buf *bytes.Buffer) {
+			buf.Write([]byte("test"))
+		}).Times(1)
+
+	mockBatch.EXPECT().Set(gomock.Any(), gomock.Any()).Return(fmt.Errorf("error")).Times(1)
+
+	err := ndb.SaveFastNode(mockNode)
+	require.Error(t, err)
+}
+
+func TestNodeDB_SaveFastNode(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	mockNode := NewMockNode(ctrl)
+	ndb, _, mockBatch, _ := createNodeDbWithMocks(ctrl)
+
+	var buf bytes.Buffer
+	buf.Grow(40)
+
+	mockNode.EXPECT().getKey().Return([]byte("k1")).Times(4)
+	mockNode.EXPECT().EncodedSize().Return(1).Times(1)
+	mockNode.EXPECT().WriteBytes(gomock.AssignableToTypeOf(&buf)).Return(nil).
+		Do(func(buf *bytes.Buffer) {
+			buf.Write([]byte("test"))
+		}).Times(1)
+
+	mockBatch.EXPECT().Set(gomock.Any(), gomock.Any()).Return(nil).Times(1)
+
+	err := ndb.SaveFastNode(mockNode)
+	require.NoError(t, err)
+}
+
+func TestNodeDB_GetNodeFailsWhenNil(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	ndb, dbMock, _, getConstructorCall := createNodeDbWithMocks(ctrl)
+	require.Panics(t, func() {
+		ndb.GetNode(nil)
+	})
+	gomock.InOrder(
+		getConstructorCall,
+		dbMock.EXPECT().Get(gomock.Any()).Return(nil, nil),
+	)
+
+	require.Panics(t, func() {
+		ndb.GetNode([]byte("k1"))
+	})
+
+}
+
+func TestNodeDB_SaveNode(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	ndb, _, mockBatch, _ := createNodeDbWithMocks(ctrl)
+	nilNode := NewMockComplexNode(ctrl)
+	nilNode.EXPECT().getHash().Return(nil).Times(1)
+	require.Panics(t, func() {
+		ndb.SaveNode(nilNode)
+	})
+
+	alreadyPersistedNode := NewMockComplexNode(ctrl)
+	alreadyPersistedNode.EXPECT().getHash().Return([]byte("Fred")).Times(1)
+	alreadyPersistedNode.EXPECT().Persisted().Return(true).Times(1)
+	require.Panics(t, func() {
+		ndb.SaveNode(alreadyPersistedNode)
+	})
+
+	happyNode := NewMockComplexNode(ctrl)
+	happyNode.EXPECT().getHash().Return([]byte("Fred")).Times(5)
+	happyNode.EXPECT().Persisted().Return(false).Times(1)
+	happyNode.EXPECT().EncodedSize().Return(10).Times(1)
+	happyNode.EXPECT().WriteBytes(gomock.Any()).Return(nil).Times(1)
+	happyNode.EXPECT().setPersisted(true).Times(1)
+
+	mockBatch.EXPECT().Set(gomock.Any(), gomock.Any()).Return(nil)
+
+	ndb.SaveNode(happyNode)
+
+}
+
+func createNodeDbWithMocks(ctrl *gomock.Controller) (*nodeDB, *mock.MockDB, *mock.MockBatch, *gomock.Call) {
```

Comment:
Could you add comments please about what this method does and what it is returning, please?